proxy.go

```diff
@@ -643,26 +643,63 @@ func (rp *reverseProxy) getUser(name string, password string) (found bool, u *us
 	defer rp.lock.RUnlock()
 	found = false
 	u = rp.users[name]
-	if u != nil {
+	switch {
+	case u != nil:
 		found = (u.password == password)
 		// existence of c and cu for toCluster is guaranteed by applyConfig
 		c = rp.clusters[u.toCluster]
 		cu = c.users[u.toUser]
-	} else if rp.hasWildcarded {
-		if s := strings.Split(name, "_"); len(s) == 2 {
-			u = rp.users[s[0]+"_*"]
-			if u != nil && u.isWildcarded {
-				c = rp.clusters[u.toCluster]
-				wildcardedCu := c.users[u.toUser]
-				if wildcardedCu != nil {
-					found = true
-					cu = wildcardedCu
-					cu.name = name
-					cu.password = password
+	case name == "" || name == defaultUser:
+		// default user can't work with the wildcarded feature for security reasons
+		found = false
+	case rp.hasWildcarded:
+		// checking if we have wildcarded users and if username matches one 3 possibles patterns
+		for _, user := range rp.users {
+			if user.isWildcarded {
+				s := strings.Split(user.name, "*")
+				// cf a validation in config.go, the names must contains either a prefix, a suffix or a wildcard
+				switch {
+				case s[0] == "" && s[1] == "":
+					// the wildcarded user is "*"
+					return getUserInformations(rp, user, name, password)
+				case s[0] == "":
+					// the wildcarded user is "*[suffix]"
+					suffix := s[1]
+					if strings.HasSuffix(name, suffix) {
+						return getUserInformations(rp, user, name, password)
+					}
+				case s[1] == "":
+					// the wildcarded user is "[prefix]*"
+					prefix := s[0]
+					if strings.HasPrefix(name, prefix) {
+						return getUserInformations(rp, user, name, password)
+					}
 				}
 			}
 		}
 	}
+	return found, u, c, cu
+}
+
+func getUserInformations(rp *reverseProxy, user *user, name string, password string) (found bool, u *user, c *cluster, cu *clusterUser) {
```

Comment:
I don't know how to better name it but initial naming hides the fact that it relates to wildcarded users. Maybe we could completely extract wildcarded users logic away to a dedicated file with possibly dedicated struct if needed? 
```suggestion
func (rp *reverseProxy) overrideWildcardedUser(user *user, name string, password string) (found bool, u *user, c *cluster, cu *clusterUser) {
```

Reply:
I found a more explictit (but longer) name: generateWildcardedUserInformation 
Regarding putting the wildcarded logic in a specfic file, I looked at the implementation and currently there are:
- 2 lines on config.go
- 37 lines of code in 2 specific function (I created a specific function like you ask) so I'm not sure it's worth creating a dedicated file. It would make sens if all the code was like that (for example if we were using a middleware pattern)